### PR TITLE
[SDL-0189] Restructuring OnResetTimeout 

### LINKED ISSUE #2422: [SDL 0189] Restructuring OnResetTimeout
Proposal: [Restructuring OnResetTimeout](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0189-Restructuring-OnResetTimeout.md)

> This proposal is to use OnResetTimeout for applicable interfaces so as to update the function and expand it to be used by a broader set of RPCs than just in _UI_ and _TTS_ interfaces.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/556

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with the following revisions: `extensionPeriod` will be renamed to `resetPeriod` and the proposal will address that HMI Integration Guidelines will need to be updated to call out that there is not currently version negotiation between HMI and Core, so older HMI implementations will not work with this new version of Core.
Lastly, it was noted that details of deprecating an HMI API will be determined upon implementation.

> The proposal .md file was updated to reflect these revisions on 7/26/18.

### PR DESCRIPTION
**Pull request by ychernysheva**
Fixes #2422
Also fixes known issue #2829

This PR is **ready** for review.

### Risk
This PR makes **major** API changes.

### Testing Plan
* Unit testing.
* ATF testing.

### Summary
Expand OnResetTimeout RPC for all interfaces.
Note: SDL will apply OnResetTimeout only in case when resetPeriod value is greater than the time remaining from the current timeout

##### Breaking Changes
* Remove deprecated functionality
* Add Reset Timeout Handler
* Create Request Controller interface

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/request_controller_impl.h [hunk 1]
```diff
@@ -47,173 +47,75 @@
...
 namespace application_manager {
 
 namespace request_controller {
 
-/**
- * @brief RequestController class is used to control currently active mobile
- * requests.
- */
-class RequestController {
+class RequestControllerImpl : public RequestController {
```
Comment (atiwari9) [inline src/components/application_manager/include/application_manager/request_controller_impl.h:60]: Please add all the deleted comments as applicable.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc [hunk 2]
```diff
@@ -45,7 +45,10 @@ RCGetInteriorVehicleDataConsentRequest::RCGetInteriorVehicleDataConsentRequest(
                                                   params.application_manager_,
                                                   params.rpc_service_,
                                                   params.hmi_capabilities_,
-                                                  params.policy_handler_) {}
+                                                  params.policy_handler_) {
```
Comment (atiwari9) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc:48]: Shouldn't this be read from smartDeviceLink.ini config file. Why is this timeout doubled?

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/button_press_request.cc [hunk 3]
```diff
@@ -54,6 +54,21 @@ ButtonPressRequest::ButtonPressRequest(
 
 ButtonPressRequest::~ButtonPressRequest() {}
 
+bool ButtonPressRequest::Init() {
```
Comment (atiwari9) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/button_press_request.cc:57]: Again, this should be HMI's responsibility. Why are we changing timeout in SDL?

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/set_interior_vehicle_data_request.cc [hunk 4]
```diff
@@ -54,6 +54,22 @@ SetInteriorVehicleDataRequest::SetInteriorVehicleDataRequest(
 
 SetInteriorVehicleDataRequest::~SetInteriorVehicleDataRequest() {}
```
Comment (atiwari9) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/set_interior_vehicle_data_request.cc:56]: Same comment as above.  this should be HMI's responsibility. Why are we changing timeout in SDL?

**Review by atiwari9 — CHANGES_REQUESTED** [review]
Thanks for the PR, i have left some review comments. Also, can you please confirm if `BC.OnResetTimeout` is applicable to all RPCs including `Alert`, `DialNumber`, `SubtleAlert` etc.? Or do we have specific requirements barring that? I noticed that there are some tests in atf test scripts designed to ignore `OnResetTimeout` for these RPCs.

File: src/components/application_manager/include/application_manager/request_controller_impl.h (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/request_controller_impl.h:60]: @atiwari9 please note that `RequestControllerImpl` is now just a class containing public functions overridden from the interface class. We don't need to duplicate the description of such functions in the impl classes according to our coding style:
https://google.github.io/styleguide/cppguide.html#Function_Comments

> When documenting function overrides, focus on the specifics of the override itself, rather than repeating the comment from the overridden function. In many of these cases, the override needs no additional documentation and thus no comment is required.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc (hunk 2 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc:48]: @atiwari9
Yes, initially default timeout value is taken from the SDL config file for all command requests:
https://github.com/smartdevicelink/sdl_core/blob/60e3354b17bcecd20d7727c71a58b34ba81f4af0/src/components/application_manager/src/commands/command_impl.cc#L84

Later, some particular commands are adjusting the default timeout according to their internal needs.
In this case, timeout is doubled to give HMI extra time for showing user consent prompt popup (which is another +10 seconds) - please look at linked issue #2829 for more details and sequence diagram.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/button_press_request.cc (hunk 3 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/button_press_request.cc:57]: @atiwari9
SDL should understand from its side whether the mobile request expired (and respond with `GENERIC_ERROR`) or HMI is still working on it. The reason for changing the timeout is the same as in the previous comment.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/set_interior_vehicle_data_request.cc (hunk 4 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/set_interior_vehicle_data_request.cc:56]: @atiwari9
SDL should understand from its side whether the mobile request expired (and respond with `GENERIC_ERROR`) or HMI is still working on it. The reason for changing the timeout is the same as in the previous comment.

**Comment by AKalinich-Luxoft** [comment]
> Thanks for the PR, i have left some review comments. Also, can you please confirm if `BC.OnResetTimeout` is applicable to all RPCs including `Alert`, `DialNumber`, `SubtleAlert` etc.? Or do we have specific requirements barring that? I noticed that there are some tests in atf test scripts designed to ignore `OnResetTimeout` for these RPCs.

@atiwari9 it's hard to simply say yes or no here. Basically, `BC.OnResetTimeout` is applicable to all RPCs until some specific requirement blocks that. For example, we have a requirement that timeout reset can't be applied to `DialNumber`. Also, we have a requirement that timeout is not applicable to `Alert`/`SubtleAlert` in case if these requests contain soft buttons and so on...

More details and references will be provided for all these requirements in atf test scripts PR

**Comment by atiwari9** [comment]
@ychernysheva  - Also can you please log a bug for the case where Alert response is still ABORTED when TTS.Alert timeout is reset and UI.Alert finishes. Ideally it should wait for both UI and TTS Alert timeouts to be complete.

**Comment by AKalinich-Luxoft** [comment]
> @ychernysheva - Also can you please log a bug for the case where Alert response is still ABORTED when TTS.Alert timeout is reset and UI.Alert finishes. Ideally it should wait for both UI and TTS Alert timeouts to be complete.

@atiwari9 I don't think this is a bug. It's a part of the existing SDL logic - in case if SDL receives a response from HMI on the UI part but the TTS part is still in progress, then SDL sends `TTS.StopSpeaking` to abort the TTS part, waits for HMI to abort TTS, and then responds with ABORTED code.
Implementation of the possibility to reset timeout for the TTS part only allows reproducing the case when TTS response is received after the UI part (however ideally TTS response should be received before UI part).
In other words, I don't think SDL should wait for both UI and TTS - response to the UI part is sufficient

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc (hunk 2 above)
Comment (atiwari9) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc:48]: I understand that this may have been used as a work around earlier, when HMI probably didn't have a way to reset the timeout on SDL side, now it does. So we really don't and shouldn't still add such patches in SDL. If this is done to maintain the legacy implementation, i understand. But even in that case please log a bug/improvement item so that the onus can now be shifted back to HMI. After all, that's the whole point of enabling HMI to reset timeout for RPCs.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/button_press_request.cc (hunk 3 above)
Comment (atiwari9) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/button_press_request.cc:57]: Please see comment: https://github.com/smartdevicelink/sdl_core/pull/3726#discussion_r658758488

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/set_interior_vehicle_data_request.cc (hunk 4 above)
Comment (atiwari9) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/set_interior_vehicle_data_request.cc:56]: https://github.com/smartdevicelink/sdl_core/pull/3726#discussion_r658758488

**Comment by atiwari9** [comment]
> > @ychernysheva - Also can you please log a bug for the case where Alert response is still ABORTED when TTS.Alert timeout is reset and UI.Alert finishes. Ideally it should wait for both UI and TTS Alert timeouts to be complete.
> 
> @atiwari9 I don't think this is a bug. It's a part of the existing SDL logic - in case if SDL receives a response from HMI on the UI part but the TTS part is still in progress, then SDL sends `TTS.StopSpeaking` to abort the TTS part, waits for HMI to abort TTS, and then responds with ABORTED code.
> Implementation of the possibility to reset timeout for the TTS part only allows reproducing the case when TTS response is received after the UI part (however ideally TTS response should be received before UI part).
> In other words, I don't think SDL should wait for both UI and TTS - response to the UI part is sufficient

Thinking further on it, i think the scenario where user has dismissed the Alert while TTS was playing does count as `ABORTED`, so it is fine. Though i am strongly for HMI to manage that instead of putting it on SDL. SDL should not determine the user experience HMI wants for its users, hence i am usually not a proponent of SDL side workarounds/patches as it creates a hard dependency. But since this is legacy implementation, we can let this particular case slide.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc (hunk 2 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_get_interior_vehicle_data_consent_request.cc:48]: @atiwari9 ok, I think we can remove all these timeout adjustments from existing PR and rely on HMI to control the lifetime of corresponding requests properly. In that case, we need the following changes:
1. Remove timeout adjustment changes from affected SDL core requests
2. Update the same requests on the HMI side - if HMI notices that extra time from the user is required then HMI sends OnResetTimeout to SDL with a doubled timeout value so SDL can wait for enough time
3. Check with ATF scripts that updated implementation does not break the legacy code and feature works as expected

Please let me know, is that plan works for you?

File: src/components/application_manager/include/application_manager/request_controller_impl.h (hunk 1 above)
Comment (atiwari9) [reply src/components/application_manager/include/application_manager/request_controller_impl.h:60]: @AKalinich-Luxoft , @ychernysheva  - Please let me know when the PR is ready to be reviewed after above mentioned changes.

**Comment by AKalinich-Luxoft** [comment]
@atiwari9 core changes according to comment https://github.com/smartdevicelink/sdl_core/pull/3726#discussion_r658758488 have been added in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/f27c548572048a81679840eed313d49345c6f0f1 and https://github.com/smartdevicelink/sdl_core/pull/3726/commits/fd7126580e50837cf0abb8270a0516d08b844728

**Review by atiwari9 — APPROVED** [review]
Aprproved.

**Comment by LitvinenkoIra** [comment]
@theresalech This PR is ready for Livio review. Thank you!

File: src/components/application_manager/src/smart_object_keys.cc [hunk 5]
```diff
@@ -454,6 +454,10 @@ const char* const rect = "rect";
 const char* const x = "x";
 const char* const y = "y";
 const char* const preferred_fps = "preferredFPS";
+
+// OnResetTimeout
+const char* const request_id = "requestID";
+const char* const timeout_period_for_reset = "resetPeriod";
```
Comment (ShobhitAd) [inline src/components/application_manager/src/smart_object_keys.cc:460]:
```suggestion
const char* const reset_period = "resetPeriod";
```
or
```suggestion
const char* const timeout_reset_period = "resetPeriod";
```
might be a better name

File: src/components/include/application_manager/application_manager.h [hunk 6]
```diff
@@ -87,7 +89,10 @@ class RPCService;
 namespace rpc_handler {
 class RPCHandler;
 }
-
+namespace request_controller {
+class RequestTimeoutHandler;
+class RequestController;
+}  // namespace request_controller
```
Comment (ShobhitAd) [inline src/components/include/application_manager/application_manager.h:95]:
```suggestion
} // namespace request_controller
```

File: src/components/interfaces/HMI_API.xml [hunk 7]
```diff
@@ -4783,6 +4783,27 @@
     </function>
     <function name="DialNumber" messagetype="response">
     </function>
+    <function name="OnResetTimeout" messagetype="notification">
+      <description>
+        HMI must send this notification to SDL for method instance for which timeout needs to be reset
+      </description>
+      <param name="requestID" type="Integer" minvalue="0" maxvalue="65535" mandatory="true">
```
Comment (ShobhitAd) [inline src/components/interfaces/HMI_API.xml:4790]: Could you please explain why the `maxvalue` is set to 65535 for this parameter

File: src/components/interfaces/HMI_API.xml [hunk 8]
```diff
@@ -4783,6 +4783,27 @@
...
+        <description>
+          Id between HMI and SDL which SDL used to send the request for method in question, for which timeout needs to be reset.
+        </description>
+      </param>
+      <param name="methodName" type="String" mandatory="true">
+        <description>
+          Name of the function for which timeout needs to be reset
+        </description>
+      </param>
+      <param name="resetPeriod" type="Integer" minvalue="0" maxvalue="1000000" mandatory="false">
```
Comment (ShobhitAd) [inline src/components/interfaces/HMI_API.xml:4800]: Could you please explain why the `maxvalue` is set to 1000000 for this parameter

**Review by ShobhitAd — CHANGES_REQUESTED** [review]

File: src/components/interfaces/HMI_API.xml (hunk 7 above)
Comment (AKalinich-Luxoft) [reply src/components/interfaces/HMI_API.xml:4790]: @ShobhitAd I think it is mostly for documentation purposes to define the border values of uint16 type. Similar to [this](https://github.com/smartdevicelink/rpc_spec/blob/d4990d8da044a0b465493449317f22ccedf949bb/MOBILE_API.xml#L1230) line as an example, I believe this is mostly a common practice to define min/max values for such parameters.

File: src/components/interfaces/HMI_API.xml (hunk 8 above)
Comment (AKalinich-Luxoft) [reply src/components/interfaces/HMI_API.xml:4800]: @ShobhitAd in this case, I believe value has been taken directly from [the proposal](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0189-Restructuring-OnResetTimeout.md#basiccommunication-onresettimeout-would-be-added).

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 9]
```diff
@@ -53,12 +53,12 @@
 #include "application_manager/hmi_interfaces_impl.h"
 #include "application_manager/message.h"
 #include "application_manager/message_helper.h"
-#include "application_manager/request_controller.h"
 #include "application_manager/resumption/resume_ctrl.h"
 #include "application_manager/rpc_handler.h"
 #include "application_manager/rpc_service.h"
 #include "application_manager/state_controller_impl.h"
 
+#include "application_manager/request_timeout_handler.h"
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @ShobhitAd removed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 10]
```diff
@@ -999,6 +999,16 @@ class ApplicationManagerImpl
     return *rpc_handler_;
   }
 
+  request_controller::RequestTimeoutHandler& get_request_timeout_handler()
+      const OVERRIDE {
+    return *request_timeout_handler_;
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/application_manager_impl.h:1004]: @ShobhitAd added in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 11]
```diff
@@ -999,6 +999,16 @@ class ApplicationManagerImpl
...
+  }
+
+  request_controller::RequestController& get_request_controller()
+      const OVERRIDE {
+    return *request_ctrl_;
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/application_manager_impl.h:1010]: @ShobhitAd added in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/include/application_manager/commands/command_request_impl.h [hunk 12]
```diff
@@ -390,6 +390,13 @@ class CommandRequestImpl : public CommandImpl,
    */
   void AddTimeOutComponentInfoToMessage(
       smart_objects::SmartObject& response) const;
+  /**
+   * @brief AddRequestToTimeoutHandler checks the request and adds it to
+   * request_timeout_handler map for tracking
+   * @param request request to HMI
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/commands/command_request_impl.h]: @ShobhitAd renamed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/include/application_manager/request_controller_impl.h [hunk 13]
```diff
@@ -47,173 +47,70 @@
 #include "interfaces/HMI_API.h"
 #include "interfaces/MOBILE_API.h"
 
+#include "application_manager/request_controller.h"
 #include "application_manager/request_controller_settings.h"
 #include "application_manager/request_info.h"
+#include "application_manager/request_timeout_handler_impl.h"
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/request_controller_impl.h]: @ShobhitAd removed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/include/application_manager/request_info.h [hunk 14]
```diff
@@ -127,7 +127,7 @@ struct RequestInfo {
   }
   uint64_t hash();
   static uint64_t GenerateHash(uint32_t var1, uint32_t var2);
-  static uint32_t HmiConnectionKey;
+  static constexpr uint32_t HmiConnectionKey = 0;
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/request_info.h]: @ShobhitAd renamed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/include/application_manager/request_timeout_handler_impl.h [hunk 15]
```diff
@@ -30,54 +30,38 @@
...
+                  const Request& request) OVERRIDE;
+  void RemoveRequest(const uint32_t hmi_correlation_id) OVERRIDE;
+  void on_event(const event_engine::Event& event) OVERRIDE;
 
  private:
-  DISALLOW_COPY_AND_ASSIGN(OnTTSResetTimeoutNotification);
-};
-
-}  // namespace hmi
+  bool IsTimeoutUpdateRequired(const Request& request,
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/include/application_manager/request_timeout_handler_impl.h:63]: @ShobhitAd added in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/application_manager/src/smart_object_keys.cc (hunk 5 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/src/smart_object_keys.cc]: @ShobhitAd renamed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

File: src/components/include/application_manager/application_manager.h (hunk 6 above)
Comment (AKalinich-Luxoft) [reply src/components/include/application_manager/application_manager.h:95]: @ShobhitAd fixed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/57e4accc52957086f89470059f8fa07d9ee5dd7c

**Comment by AKalinich-Luxoft** [comment]
@ShobhitAd all comments have been processed. Could you please proceed with your review here?

**Review by ShobhitAd — APPROVED** [review]

**Comment by iCollin** [comment]
@AKalinich-Luxoft Hello,

With these changes `OnResetTimeout` with no `resetPeriod` specified will set the timeout equal to the `defaultTimeout` from the ini. Previously, the HMI requesting a ResetTimeout would set the timeout to a command's modified `default_timeout_` property, which is greater than `defaultTimeout`. Because of this change, if the HMIs now send a `TIMEOUT` message after the `defaultTimeout`, Core will have sent mobile a `GENERIC_ERROR` response before the HMI's response can be processed, resulting in Mobile not receiving responses it should. 

In order to handle the time the message spends on the wire (in transport), in a message queue, and finally being processed I propose we add some additional time to these timers, ideally configurable from the INI.

At the moment these `TIMEOUT` responses can be received by mobile because the HMI is decrementing all incoming timeouts by 1000ms, which will be removed.

**Comment by AKalinich-Luxoft** [comment]
> @AKalinich-Luxoft Hello,
> 
> With these changes `OnResetTimeout` with no `resetPeriod` specified will set the timeout equal to the `defaultTimeout` from the ini. Previously, the HMI requesting a ResetTimeout would set the timeout to a command's modified `default_timeout_` property, which is greater than `defaultTimeout`. Because of this change, if the HMIs now send a `TIMEOUT` message after the `defaultTimeout`, Core will have sent mobile a `GENERIC_ERROR` response before the HMI's response can be processed, resulting in Mobile not receiving responses it should.
> 
> In order to handle the time the message spends on the wire (in transport), in a message queue, and finally being processed I propose we add some additional time to these timers, ideally configurable from the INI.
> 
> At the moment these `TIMEOUT` responses can be received by mobile because the HMI is decrementing all incoming timeouts by 1000ms, which will be removed.

@iCollin all changes have been implemented in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/031c65f71660535ba75e9513c433a40e3871d48a
We added a default timeout compensation param to ini file which is defaulted to 1000ms.
This timeout compensation is applied to all commands timeouts initially and also applied after OnResetTimeout notification.
All affected unit tests were updated accordingly.

File: src/components/config_profile/include/config_profile/profile.h [hunk 16]
```diff
@@ -249,6 +249,12 @@ class Profile : public protocol_handler::ProtocolHandlerSettings,
    */
   const uint32_t& default_timeout() const;
 
+  /**
+   * @brief Default timeout compensation for waiting for response to mobile app
+   * some extra time
```
Comment (ShobhitAd) [inline src/components/config_profile/include/config_profile/profile.h:254]:
```suggestion
   * @brief Default timeout compensation for waiting some extra time for response to mobile app
```

**Review by ShobhitAd — CHANGES_REQUESTED** [review]

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc [hunk 17]
```diff
@@ -278,7 +278,13 @@ void RCCommandRequest::ProcessConsentResult(const bool is_allowed,
   SDL_LOG_AUTO_TRACE();
   if (is_allowed) {
     SetResourceState(module_type, ResourceState::BUSY);
+    const auto default_timeout =
+        application_manager_.get_settings().default_timeout() +
+        application_manager_.get_settings().default_timeout_compensation();
+    application_manager_.UpdateRequestTimeout(
+        connection_key(), correlation_id(), default_timeout);
     Execute();  // run child's logic
+
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc]: @ShobhitAd fixed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/55d750086399afee1b669105e487d7a4611e2dd6

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/navi_start_stream_request.cc [hunk 18]
```diff
@@ -58,9 +58,11 @@ NaviStartStreamRequest::NaviStartStreamRequest(
   SDL_LOG_AUTO_TRACE();
   std::pair<uint32_t, int32_t> stream_retry =
       application_manager_.get_settings().start_stream_retry_amount();
-  default_timeout_ = stream_retry.second;
+  default_timeout_ =
+      stream_retry.second +
+      application_manager_.get_settings().default_timeout_compensation();
   retry_number_ = stream_retry.first;
-  SDL_LOG_DEBUG("default_timeout_ = " << default_timeout_
+  SDL_LOG_DEBUG("default_timeout_ = " << default_timeout()
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/navi_start_stream_request.cc]: @ShobhitAd fixed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/55d750086399afee1b669105e487d7a4611e2dd6

[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/navi_start_stream_request.cc]: As well as in audio streaming request

File: src/components/config_profile/include/config_profile/profile.h (hunk 16 above)
Comment (AKalinich-Luxoft) [reply src/components/config_profile/include/config_profile/profile.h]: @ShobhitAd fixed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/55d750086399afee1b669105e487d7a4611e2dd6

**Comment by ShobhitAd** [comment]
@AKalinich-Luxoft thank you for addressing the review comments. Please resolve the merge conflicts on the PR

**Comment by AKalinich-Luxoft** [comment]
> @AKalinich-Luxoft thank you for addressing the review comments. Please resolve the merge conflicts on the PR

@ShobhitAd done

**Comment by ShobhitAd** [comment]
@AKalinich-Luxoft a small comment regarding the rpc_spec commit change.
It looks like the rpc_spec commit used in the PR is [rpc_spec @ d4990d8](https://github.com/smartdevicelink/rpc_spec/tree/d4990d8da044a0b465493449317f22ccedf949bb) but the latest rpc_spec commit for develop is [rpc_spec @ 12c2d1d](https://github.com/smartdevicelink/rpc_spec/commit/12c2d1dacc4315be22bbc1f5e655f4464d9fae44
Could you update the rpc_spec submodule commit in the PR to point to the latest commit on rpc_spec develop

**Comment by AKalinich-Luxoft** [comment]
@ShobhitAd fixed in https://github.com/smartdevicelink/sdl_core/pull/3726/commits/c011a9c31c9c33db424ba1f8c565de3f1b9053eb
Sorry, forgot to fetch this repo

**Review by ShobhitAd — APPROVED** [review]
